README.md

```diff
@@ -0,0 +1,110 @@
+# WebSocket - Go
+
+A simple [WebSocket](https://datatracker.ietf.org/doc/html/rfc6455) server that performs the HTTP upgrade and prints log messages on all standardized WebSocket events, such as `open`, `message`, `close` and `error`. The server is written in Golang and uses the [Gorilla WebSocket](github.com/gorilla/websocket) library.
+
+
+## Before you begin
+
+- A Kubernetes cluster with Knative installed and DNS configured. See
+  [Install Knative Serving](https://knative.dev/docs/install/serving/install-serving-with-yaml).
+- [ko](https://github.com/ko-build/ko) installed and running on your local machine,
+  and a Docker Hub account configured (we'll use it for a container registry).
+
+## The sample code.
+
+1. If you look in `cmd/server/main.go`, you will the `main` function setting a `handleWebSocket` function and starting the web server on the `/ws` context:
+
+   ```go
+    func main() {
+        http.HandleFunc("/ws", handleWebSocket)
+        fmt.Println("Starting server on :8080...")
+        if err := http.ListenAndServe(":8080", nil); err != nil {
+            log.Fatalf("Server error: %v", err)
+        }
+    }
+   ```
+
+2. The `handleWebSocket` performs the protocol upgrade and assigns various websocket handler functions, such as `OnOpen` or `OnMessage`:
+
+   ```go
+    func handleWebSocket(w http.ResponseWriter, r *http.Request) {
+        conn, err := upgrader.Upgrade(w, r, nil)
+        if err != nil {
+            log.Printf("Error upgrading to websocket: %v", err)
+            return
+        }
+        handlers.OnOpen(conn)
+
+        go func() {
+            defer handlers.OnClose(conn)
+            for {
+                messageType, message, err := conn.ReadMessage()
+                if err != nil {
+                    handlers.OnError(conn, err)
+                    break
+                }
+                handlers.OnMessage(conn, messageType, message)
+            }
+        }()
+    }
+   ```
+
+3. The WebSocket application logic is located in the `pkg/handlers/handlers.go` file and contains callbacks for each WebSocket event:
+
+   ```go
+    func OnOpen(conn *websocket.Conn) {
+        log.Printf("WebSocket connection opened: %v", conn.RemoteAddr())
+    }
+
+    func OnMessage(conn *websocket.Conn, messageType int, message []byte) {
+        log.Printf("Received message from %v: %s", conn.RemoteAddr(), string(message))
+
+        if err := conn.WriteMessage(messageType, message); err != nil {
+            log.Printf("Error sending message: %v", err)
+        }
+    }
+
+    func OnClose(conn *websocket.Conn) {
+        log.Printf("WebSocket connection closed: %v", conn.RemoteAddr())
+        conn.Close()
+    }
+
+    func OnError(conn *websocket.Conn, err error) {
+        log.Printf("WebSocket error from %v: %v", conn.RemoteAddr(), err)
+    }    
+   ```
+
+## Build and Deploy the application
+
+At the the root of directory run the following command:
+
+   ```bash
+   ko apply -f config/service.yaml
```

Comment:
In the samples we use `docker buildx build --platform linux/arm64,linux/amd64 -t "{username}/grpc-ping-go" --push .`